src/test.js

```diff
@@ -158,8 +158,6 @@ Test.prototype = {
     return moduleStartChain.then(() => {
       config.current = this;
 
-      this.testEnvironment = extend({}, module.testEnvironment);
```

Comment:
with the current approach nothing should access this until the `before` hooks have completed

Reply:
OK. In that case we'll need to keep in mind to note in the 3.0 Upgrade Guide that `QUnit.moduleStart()` callbacks can no longer access testEnvironment via [`QUnit.config.current`](https://qunitjs.com/api/config/current/), and recommend `QUnit.hooks.before` for that use case instead.

I'm not aware of this being common, so that's fine I think and could be worth doing. Perhaps more likely, is access to `QUnit.config.current` from `QUnit.testStart()`. It's ambigious whether that intends to modify the base context for all tests in the current module during the first test. Most likely not intended. But also most likely harmless since it'll re-run for each test. Making that no longer modify the base context could be seen as a bug fix and either way, a change I think is good to make.

But... I think the current version of the PR does more than that, it makes it undefined. That too could be mitigated by recommending use of `QUnit.hooks.beforeEach` instead in QUnit 3.0, but I'm less sure whether that's worth breaking. What do you think?

Reply:
so if we set `this.testEnvironment` to the module context here, conditional modifications to `QUnit.config.current.testEnvironment` would modify the module context and could "leak" awkwardly and impact following tests in unintended ways.

we could set `this.testEnvironment` to `extend({}, module.testEnvironment, false, true)` and clobber it when it gets set again later, which would allow reading it during `testStart` but not allow modifying it.

keeping it this way means it is `undefined` during `testStart` and can't be read or modified.

another option could be to make a sort of implied single-test module context here. so we'd do
```
this.testEnvironment = Object.create(module.testEnvironment || {});
```
here and use the resulting `this.testEnvironment` for the `extend` call in the test queue later on, so the test hooks still receive a flattened copy. Which I think would preserve modifications made during `testStart` _and_ only impact the current test? 

If the last options sounds worth doing then maybe we should add test coverage on that behavior and I can rebase with that strategy to make sure it still works how I'm imagining.